Import MyDataBase class.

In [1]:
from src.database import MyDataBase
db_name = "parch_and_posey"

Load database.

In [2]:
database = MyDataBase(db_name)
database.print_tables_names()

ISOLATION_LEVEL_AUTOCOMMIT: 0
('web_events',)
('sales_reps',)
('region',)
('orders',)
('accounts',)


### LEFT & RIGHT Quizzes

1. In the accounts table, there is a column holding the website for each company.
The last three digits specify what type of web address they are using.
A list of extensions (and pricing) is provided [here](https://iwantmyname.com/domains).
Pull these extensions and provide how many of each website type exist in the accounts table.

In [3]:
q = """
SELECT RIGHT(website, 3) AS domain, COUNT(*) num_companies
FROM accounts
GROUP BY 1
ORDER BY 2 DESC;
"""

table = database.to_data_frame(q)
table

domain  num_companies
0    com            349
1    net              1
2    org              1

2. There is much debate about how much the name (or even the first letter of a company name) matters.
Use the accounts table to pull the first letter of each company name
to see the distribution of company names that begin with each letter (or number).

In [4]:
q = """
SELECT LEFT(UPPER(name), 1) AS first_letter, COUNT(*) num_companies
FROM accounts
GROUP BY 1
ORDER BY 2 DESC;
"""

table = database.to_data_frame(q)
table

first_letter  num_companies
0             C             37
1             A             37
2             P             27
3             M             22
4             S             17
5             D             17
6             T             17
7             B             16
8             E             16
9             L             16
10            H             15
11            N             15
12            G             14
13            U             13
14            W             12
15            F             12
16            R              8
17            J              7
18            K              7
19            O              7
20            I              7
21            V              7
22            X              2
23            3              1
24            Y              1
25            Q              1

3. Use the accounts table and a CASE statement to create two groups:
one group of company names that start with a number
and a second group of those company names that start with a letter.
What proportion of company names start with a letter?

In [5]:
q = """
SELECT SUM(num) nums, SUM(letter) letters
FROM (SELECT name, CASE WHEN LEFT(UPPER(name), 1) IN ('0','1','2','3','4','5','6','7','8','9')
                       THEN 1 ELSE 0 END AS num,
         CASE WHEN LEFT(UPPER(name), 1) IN ('0','1','2','3','4','5','6','7','8','9')
                       THEN 0 ELSE 1 END AS letter
      FROM accounts) t1;
"""

table = database.to_data_frame(q)
table

nums  letters
0     1      350

There are 350 company names that start with a letter and 1 that starts with a number. This gives a ratio of 350/351 that are company names that start with a letter or 99.7%.

4. Consider vowels as a, e, i, o, and u.
What proportion of company names start with a vowel, and what percent start with anything else?

In [6]:
q = """
SELECT SUM(vowels) vowels, SUM(other) other
FROM (SELECT name, CASE WHEN LEFT(UPPER(name), 1) IN ('A','E','I','O','U')
                        THEN 1 ELSE 0 END AS vowels,
          CASE WHEN LEFT(UPPER(name), 1) IN ('A','E','I','O','U')
                       THEN 0 ELSE 1 END AS other
         FROM accounts) t1;
"""

table = database.to_data_frame(q)
table

vowels  other
0      80    271

There are 80 company names that start with a vowel and 271 that start with other characters.
Therefore 80/351 are vowels or 22.8%. Therefore, 77.2% of company names do not start with vowels.


### Quizzes POSITION & STRPOS


1. Use the accounts table to create first and last name columns
that hold the first and last names for the primary_poc.

In [7]:
q = """
SELECT LEFT(primary_poc, STRPOS(primary_poc, ' ') -1 ) first_name,
RIGHT(primary_poc, LENGTH(primary_poc) - STRPOS(primary_poc, ' ')) last_name
FROM accounts;
"""

table = database.to_data_frame(q)
table

first_name   last_name
0       Tamara        Tuma
1         Sung     Shields
2        Jodee        Lupo
3     Serafina       Banda
4      Angeles      Crusoe
..         ...         ...
346      Buffy       Azure
347       Esta  Engelhardt
348   Khadijah     Riemann
349     Deanne    Hertlein
350     Jewell       Likes

[351 rows x 2 columns]

2. Now see if you can do the same thing for every rep name in the sales_reps table.
Again provide first and last name columns.

In [8]:
q = """
SELECT LEFT(name, STRPOS(name, ' ') -1 ) first_name,
       RIGHT(name, LENGTH(name) - STRPOS(name, ' ')) last_name
FROM sales_reps;
"""

table = database.to_data_frame(q)
table

first_name     last_name
0       Samuel        Racine
1       Eugena         Esser
2       Michel      Averette
3      Renetta         Carew
4         Cara        Clarke
5       Lavera          Oles
6         Elba        Felder
7     Shawanda         Selke
8        Sibyl        Lauria
9       Necole       Victory
10   Ernestine       Pickron
11      Ayesha        Monica
12       Retha         Sears
13       Julia       Behrman
14         Tia         Amato
15      Akilah      Drinkard
16     Silvana        Virden
17     Nakesha          Renn
18        Elna      Condello
19      Gianna        Dossey
20     Debroah        Wardle
21    Sherlene  Wetherington
22        Chau        Rowles
23    Carletta      Kosinski
24     Charles       Bidwell
25       Cliff        Meints
26     Delilah          Krum
27    Kathleen       Lalonde
28       Julie         Starr
29     Cordell        Rieder
30      Earlie    Schleusner
31        Moon        Torian
32     Dorotha       Seawell
33       Maren         Musto
34     Vernita         Plump
35      Calvin       Ollison
36       Saran           Ram
37     Derrick       Boggess
38     Babette        Soukup
39       Nelle         Meaux
40      Soraya        Fulton
41     Brandie          Riva
42   Marquetta       Laycock
43       Hilma        Busick
44       Arica     Stoltzfus
45      Elwood         Shutt
46    Maryanna    Fiorentino
47  Georgianna      Chisholm
48       Micha      Woodford
49       Dawna         Agnew